Enforce that test classes start with "Test" instead of suffix

CuteChuanChuan opened the pull request:
Summary:

Closes https://github.com/apache/iceberg/issues/13442

Details:

- Add checkstyle rule to enforce test class naming conventions, using 'Test' as prefix
- Rename test classes to comply with the naming conventions

Verification:

1. Add test cases locally to ensure checkstyle functionality and verify the new rule works as expected
2. Run `./gradlew clean build -x integrationTest --no-build-cache` to verify modifications don't break existing functionality and ensure no ErrorProne warnings

Appendix:
- Script to add fake test classes, run checkstyle, and remove these classes:
```
#!/bin/bash

# Test Case 1: TestMyCustomExample (should NOT trigger - starts with Test)
cat > core/src/test/java/org/apache/iceberg/TestMyCustomExample.java << 'EOF'
package org.apache.iceberg;

public class TestMyCustomExample {
    // Test class starting with Test - should NOT trigger violation
}
EOF

# Test Case 2: MyUtilsTest (should trigger - ends with Test)
cat > core/src/test/java/org/apache/iceberg/MyUtilsTest.java << 'EOF'
package org.apache.iceberg;

public class MyUtilsTest {
    // Utils test class ending with Test - should trigger violation
}
EOF

# Test Case 3: MyTestBase (should NOT trigger - base class, no Test suffix)
cat > core/src/test/java/org/apache/iceberg/MyTestBase.java << 'EOF'
package org.apache.iceberg;

public abstract class MyTestBase {
    // Abstract test base class - should NOT trigger violation (no Test suffix)
}
EOF

# Run checkstyle
./gradlew :iceberg-core:checkstyleTest

# Clean up
rm -f core/src/test/java/org/apache/iceberg/TestMyCustomExample.java
rm -f core/src/test/java/org/apache/iceberg/MyUtilsTest.java
rm -f core/src/test/java/org/apache/iceberg/MyTestBase.java
```

nastra commented on core/src/test/java/org/apache/iceberg/data/TestData.java:
```diff
@@ -53,7 +53,7 @@
 import org.junit.jupiter.params.provider.MethodSource;
 
-public abstract class DataTest {
+public abstract class TestData {
```
let's call this `DataTestBase` since it's a base class for a bunch of other tests

CuteChuanChuan replied:
Thanks for the suggestions!

nastra commented on data/src/test/java/org/apache/iceberg/TestRecordWrapper.java:
```diff
@@ -26,7 +26,7 @@
 import org.junit.jupiter.api.Test;
 
-public abstract class RecordWrapperTest {
+public abstract class TestRecordWrapper {
```
```suggestion
public abstract class RecordWrapperTestBase {
```

nastra commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/data/TestAvroData.java:
```diff
@@ -54,7 +54,7 @@
 import org.junit.jupiter.params.provider.MethodSource;
 
-public abstract class AvroDataTest {
+public abstract class TestAvroData {
```
```suggestion
public abstract class AvroDataTestBase {
```

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/TestAvroData.java:
```diff
@@ -54,7 +54,7 @@
 import org.junit.jupiter.params.provider.MethodSource;
 
-public abstract class AvroDataTest {
+public abstract class TestAvroData {
```
```suggestion
public abstract class AvroDataTestBase {
```

CuteChuanChuan commented on .baseline/checkstyle/checkstyle.xml:
```diff
@@ -431,6 +431,11 @@
             <property name="format" value="@Test\(.*expected.*\)"/>
             <property name="message" value="Prefer using Assertions.assertThatThrownBy(...).isInstanceOf(...) instead."/>
         </module>
+        <module name="RegexpSinglelineJava">
```
Thanks for the feedback!

CuteChuanChuan commented:
Hi @nastra ,

Thanks for your review! I've addressed all the feedback points:

✅ Added suppression rule to limit TestClassNamingConvention to test directories only
✅ Renamed abstract base classes to use *TestBase suffix pattern (DataTestBase, RecordWrapperTestBase, AvroDataTestBase)
✅ Updated all references and imports in extending classes

Appendix:
- Script to test that checkstyle is only applied to test directories:
```
#!/bin/bash

# Test Case 1: Should be ignored
cat > core/src/main/java/org/apache/iceberg/SomethingTest.java << 'EOF'
package org.apache.iceberg;

public class SomethingTest {
    // This should be ignored by TestClassNamingConvention (in main directory)
}
EOF

# Test Case 2: Should be checked
cat > core/src/test/java/org/apache/iceberg/AnotherBadTest.java << 'EOF'
package org.apache.iceberg;

public class AnotherBadTest {
    // This should trigger TestClassNamingConvention violation (in test directory)
}
EOF

# Run checkstyle
./gradlew :iceberg-core:checkstyleMain
./gradlew :iceberg-core:checkstyleTest

# Clean up
rm -f core/src/main/java/org/apache/iceberg/SomethingTest.java
rm -f core/src/test/java/org/apache/iceberg/AnotherBadTest.java
```

amogh-jahagirdar commented on spark/v4.0/spark-runtime/src/integration/java/org/apache/iceberg/spark/TestSmoke.java:
```diff
@@ -31,7 +31,7 @@
 
 @ExtendWith(ParameterizedTestExtension.class)
-public class SmokeTest extends ExtensionsTestBase {
+public class TestSmoke extends ExtensionsTestBase {
```
I think we should find a better name for this, `TestSmoke` doesn't make a lot of sense; how about `TestRoundTrip`

CuteChuanChuan replied:
Thanks for pointing this out!

CuteChuanChuan commented:
Hi @amogh-jahagirdar ,

as suggested, the test class (`TestSmoke`) was renamed. PTAL. Thanks!
